R/server.R

```diff
@@ -802,21 +830,41 @@ runExample <- function(example=NA,
 #'
 #' @export
 runGadget <- function(app, server = NULL, port = getOption("shiny.port"),
-  viewer = paneViewer()) {
+  viewer = paneViewer(), stopOnCancel = TRUE) {
 
   if (!is.shiny.appobj(app)) {
     app <- shinyApp(app, server)
   }
 
+  if (isTRUE(handleCancel)) {
+    app <- decorateServerFunc(app, function(input, output, session) {
+      observeEvent(input$cancel, {
+        stopApp(stop("User cancel"))
+      })
+    })
+  }
+
   if (is.null(viewer)) {
     viewer <- browseURL
   }
 
-  retVal <- withVisible(shiny::runApp(app, port = port, launch.browser = viewer))
-  if (retVal$visible)
-    retVal$value
-  else
-    invisible(retVal$value)
+  shiny::runApp(app, port = port, launch.browser = viewer)
+}
+
+# Add custom functionality to a Shiny app object's server func
+decorateServerFunc <- function(appobj, serverFunc) {
+  origServerFuncSource <- appobj$serverFuncSource
+  appobj$serverFuncSource <- function() {
+    origServerFunc <- origServerFuncSource()
+    function(input, output, session) {
+      serverFunc(input, output, session)
+      if (length(formals(plot)) == 2)
```

Comment:
Also, when we get to this line, has the user's server function already been checked for the number and name of arguments? I have a vague memory of the argument check happening when the app's server function is invoked, but I could be wrong. Even if it has already been checked for arguments, it would be helpful to have a comment about it here.